tools/webpack/shared.js

```diff
@@ -0,0 +1,109 @@
+/**
+ * External dependencies
+ */
+const { BundleAnalyzerPlugin } = require( 'webpack-bundle-analyzer' );
+const { DefinePlugin } = require( 'webpack' );
+const TerserPlugin = require( 'terser-webpack-plugin' );
+const { compact } = require( 'lodash' );
+const postcss = require( 'postcss' );
+
+/**
+ * WordPress dependencies
+ */
+const DependencyExtractionWebpackPlugin = require( '@wordpress/dependency-extraction-webpack-plugin' );
+const ReadableJsAssetsWebpackPlugin = require( '@wordpress/readable-js-assets-webpack-plugin' );
+
+const {
+	NODE_ENV: mode = 'development',
+	WP_DEVTOOL: devtool = mode === 'production' ? false : 'source-map',
+} = process.env;
+
+const stylesTransform = ( content ) => {
+	if ( mode === 'production' ) {
+		return postcss( [
+			require( 'cssnano' )( {
+				preset: [
+					'default',
+					{
+						discardComments: {
+							removeAll: true,
+						},
+					},
+				],
+			} ),
+		] )
+			.process( content, {
+				from: 'src/app.css',
+				to: 'dest/app.css',
+			} )
+			.then( ( result ) => result.css );
+	}
+	return content;
+};
+
+const optimization = {
+	// Only concatenate modules in production, when not analyzing bundles.
+	concatenateModules:
+		mode === 'production' && ! process.env.WP_BUNDLE_ANALYZER,
+	minimizer: [
+		new TerserPlugin( {
+			cache: true,
+			parallel: true,
+			sourceMap: mode !== 'production',
+			terserOptions: {
+				output: {
+					comments: /translators:/i,
+				},
+				compress: {
+					passes: 2,
+				},
+				mangle: {
+					reserved: [ '__', '_n', '_nx', '_x' ],
+				},
+			},
+			extractComments: false,
+		} ),
+	],
+};
+
+const moduleConfig = {
+	rules: compact( [
+		mode !== 'production' && {
+			test: /\.js$/,
+			use: require.resolve( 'source-map-loader' ),
+			enforce: 'pre',
+		},
+	] ),
+};
+
+const plugins = [
+	// The WP_BUNDLE_ANALYZER global variable enables a utility that represents bundle
+	// content as a convenient interactive zoomable treemap.
+	process.env.WP_BUNDLE_ANALYZER && new BundleAnalyzerPlugin(),
+	new DefinePlugin( {
+		// Inject the `GUTENBERG_PHASE` global, used for feature flagging.
+		'process.env.GUTENBERG_PHASE': JSON.stringify(
+			parseInt( process.env.npm_package_config_GUTENBERG_PHASE, 10 ) || 1
+		),
+		'process.env.FORCE_REDUCED_MOTION': JSON.stringify(
+			process.env.FORCE_REDUCED_MOTION
+		),
+	} ),
+	new DependencyExtractionWebpackPlugin( { injectPolyfill: true } ),
+	mode === 'production' && new ReadableJsAssetsWebpackPlugin(),
+];
+
+const watchOptions = {
+	ignored: [ '**/node_modules', '**/packages/*/src' ],
+	aggregateTimeout: 500,
+};
+
+module.exports = {
```

Comment:
It's more a nitpick: I kinda want this to export a `baseConfig` and some other utils rather than just a collection of variables here. Benefit is that it's clear that we're extending a base config and we can hide all those related variables (`devtool`, `mode`, etc) inside it. If we want to add some shared settings we can easily do that by adding them to that config too. It's also clearer to the consumer configs that this shared file didn't do anything special with those settings.

Reply:
What I mean is something like

```js
// webpack/shared.js
module.exports = {
  baseConfig: {
    // mode, devtool, etc
  },
  // styleTransform,
}
```

So that for consumer configs the can just...

```js
// webpack/packages.config.js
const { baseConfig, styleTransform } = require('./shared');

module.exports = {
  ...baseConfig,
  // ...
}
```

1. For the consumers, they can easily know whether a field is webpack-related or just some helpers we exported.
2. Adding a field to every config only have to update one place.

It's not a big deal though, it's just one or more clicks away to understand the code. It just feels more natural to me to know that I'm _extending_ a config, not _creating_ a new one. It's just my preference, feel free to ignore this :).

Reply:
https://github.com/WordPress/gutenberg/pull/33293/commits/a674d333ec261ae2dacd5fcdc3d04f7705a39a1e – this should do the trick. I left `plugins` as a separate export to avoid nested spread operator usage.

Reply:
LGTM 👍